### PR TITLE
Don't unpin MemoryPoolSlab while in use

### PR DESCRIPTION
**Pull request by benaadams**
May cause heap corruption re: https://github.com/aspnet/Common/pull/332#discussion_r183229246

Currently it Unpins, then a block is returned which throws an ODE exception https://github.com/aspnet/KestrelHttpServer/issues/2486
```
Log Critical[0]: Unexpected exception in HttpConnection.ProcessRequestsAsync. System.ObjectDisposedException: Cannot access a disposed object.
[20:27:03]	Object name: 'MemoryPoolBlock'.
[20:27:03]	at System.Buffers.ThrowHelper.ThrowObjectDisposedException(ExceptionArgument argument) in D:\b\w\1e8cd6334ef22651\packages\microsoft.extensions.buffers.sources\2.1.0-preview3-32110\contentFiles\cs\netstandard1.0\Microsoft.Extensions.Buffers.Sources\PipelinesThrowHelper.cs:line 51
[20:27:03]	at System.Buffers.MemoryPoolBlock.Dispose() in D:\b\w\1e8cd6334ef22651\packages\microsoft.extensions.buffers.sources\2.1.0-preview3-32110\contentFiles\cs\netstandard1.0\Microsoft.Extensions.Buffers.Sources\MemoryPoolBlock.Release.cs:line 54
[20:27:03]	at System.IO.Pipelines.BufferSegment.ResetMemory()
[20:27:03]	at System.IO.Pipelines.Pipe.CompletePipe()
[20:27:03]	at System.IO.Pipelines.Pipe.CompleteWriter(Exception exception)
[20:27:03]	at Microsoft.AspNetCore.Server.Kestrel.Core.Adapter.Internal.AdaptedPipeline.ReadInputAsync(Stream stream)
[20:27:03]	at Microsoft.AspNetCore.Server.Kestrel.Core.Adapter.Internal.AdaptedPipeline.RunAsync(Stream stream) in D:\b\w\1e8cd6334ef22651\src\Kestrel.Core\Adapter\Internal\AdaptedPipeline.cs:line 45
[20:27:03]	at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.HttpConnection.ProcessRequestsAsync[TContext](IHttpApplication`1 httpApplication) in D:\b\w\1e8cd6334ef22651\src\Kestrel.Core\Internal\HttpConnection.cs:line 183
``` 
Probably bad things also happen as the data block would now be using unpinned memory, with a long previously captured pointer... Like heap corruption? https://github.com/aspnet/KestrelHttpServer/issues/2503 https://github.com/dotnet/coreclr/issues/17679

New spin on this; is the change https://github.com/aspnet/Common/pull/332 started throwing ODE exceptions, whereas previously it didn't do, so that may have been where the GC crash leading to hangs surfaced in the error handling/unwind?

Resolves https://github.com/aspnet/KestrelHttpServer/issues/2486 (Test timeout: MemoryPoolBlock ObjectDisposedException)

/cc @pakrym @davidfowl @GrabYourPitchforks @mikeharder @jkotas

### CODE REVIEW

**Comment by benaadams** [comment]
Think it mostly arises in Kestrel tests; as a shutdown race

**Comment by pakrym** [comment]
I feel like we need to fix kestrel so it doesn't dispose while in use.

File: shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolBlock.Debug.cs [hunk 1]
```diff
@@ -74,14 +74,19 @@ public override Memory<byte> Memory
...
         {
-            if (!Slab.IsActive) ThrowHelper.ThrowObjectDisposedException(ExceptionArgument.MemoryPoolBlock);
-
             if (Volatile.Read(ref _pinCount) > 0)
             {
                 ThrowHelper.ThrowInvalidOperationException_ReturningPinnedBlock();
             }
 
-            Pool.Return(this);
+            if (Slab.IsActive)
```
Comment (pakrym) [inline shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolBlock.Debug.cs]: Move this logic into Slab.Return.

File: shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs [hunk 2]
```diff
@@ -66,15 +76,36 @@ protected void Dispose(bool disposing)
...
+        internal void Return(MemoryPoolBlock block)
+        {
+#if BLOCK_LEASE_TRACKING
+            Debug.Assert(block.Slab == this, "Returned block was not this slab");
+            Debug.Assert(block.IsLeased, $"Block being returned to pool twice: {block.Leaser}{Environment.NewLine}");
+            block.IsLeased = false;
+#endif
+            GC.SuppressFinalize(block);
+
+            var remaining = Interlocked.Decrement(ref _ownedBlocks);
```
Comment (pakrym) [inline shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs]: I'm concerned about leaking pinned arrays. Are you sure we would always end up inpinning them?

File: shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs [hunk 3]
```diff
@@ -66,15 +76,36 @@ protected void Dispose(bool disposing)
...
+
+            if (remaining < 0)
```
Comment (pakrym) [inline shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs]: This case is double return. To check for actual dispose use `_disposedValue`. Can we throw only when the block we are turning we double returned?

Comment (pakrym) [reply shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs]: We do with assert. Should we change assert to exception? What's easier to debug?

File: shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs (hunk 2 above)
Comment (benaadams) [reply shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs]: Not if the block was never returned

File: shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs [hunk 4]
```diff
@@ -77,6 +77,12 @@ protected void Dispose(bool disposing)
                 _isActive = false;
 
                 _disposedValue = true;
+#if DEBUG
+                if (_ownedBlocks != 0)
+                {
+                    throw new InvalidOperationException("MemoryPoolBlocks still outstanding!");
```
Comment (pakrym) [inline shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs]: Add blocks `Leaser`s to exception message.

**Comment by pakrym** [comment]
Tests for things that are reliable testable.

**Comment by benaadams** [comment]
@pakrym I'm happy for you to take this on 😉

**Comment by mikeharder** [comment]
For the record, https://github.com/aspnet/KestrelHttpServer/issues/2503 was caused by a bug in the GC (https://github.com/dotnet/coreclr/issues/17716) rather than managed heap corruption.

**Comment by davidfowl** [comment]
Closing this as there are tons of conflicts and we aren't particularly convinced this is required.

**Pull request closed without merging**